libs/constrained-generators/testlib/Test/Min/Base.hs

```diff
@@ -0,0 +1,501 @@
+-- ===========================================
+
+-- | Syntactic operations are ones that have to do with the structure and appearence of the type.
+class Syntax (t :: [Type] -> Type -> Type) where
+  inFix :: forall dom rng. t dom rng -> Bool
+  inFix _ = False
+  name :: forall dom rng. t dom rng -> String
+
+-- | Semantic operations are ones that give the function symbol, meaning as a function.
+--   I.e. how to apply the function to a list of arguments and return a value,
+--   or to apply meaning preserving rewrite rules.
+class Syntax t => Semantics (t :: [Type] -> Type -> Type) where
+  semantics :: forall d r. t d r -> FunTy d r -- e.g. FunTy '[a,Int] Bool == a -> Int -> Bool
+  rewriteRules ::
+    forall ds rng.
+    (TypeList ds, Typeable ds, HasSpec rng, All HasSpec ds) =>
+    t ds rng -> List Term ds -> Evidence (Typeable rng, Eq rng, Show rng) -> Maybe (Term rng)
+  rewriteRules t l Evidence = Lit <$> (applyFunSym @t (semantics t) l)
+
+-- | Logical operations are one that support reasoning about how a function symbol
+--   relates to logical properties, that we call Spec's
+class (Typeable t, Syntax t, Semantics t) => Logic (t :: [Type] -> Type -> Type) where
+  {-# MINIMAL propagate | (propagateTypeSpec, propagateMemberSpec) #-}
+
+  propagateTypeSpec ::
+    (AppRequires t as b, HasSpec a) =>
+    t as b -> ListCtx as (HOLE a) -> TypeSpec b -> [b] -> Spec a
+  propagateTypeSpec f ctx ts cant = propagate f ctx (TypeSpec ts cant)
+
+  propagateMemberSpec ::
+    (AppRequires t as b, HasSpec a) =>
+    t as b -> ListCtx as (HOLE a) -> NonEmpty b -> Spec a
+  propagateMemberSpec f ctx xs = propagate f ctx (MemberSpec xs)
+
+  propagate ::
+    (AppRequires t as b, HasSpec a) =>
+    t as b -> ListCtx as (HOLE a) -> Spec b -> Spec a
+  propagate _ _ TrueSpec = TrueSpec
+  propagate _ _ (ErrorSpec es) = ErrorSpec es
+  propagate f ctx (SuspendedSpec v ps) = constrained $ \v' -> Let (App f (fromListCtx ctx v')) (v :-> ps)
+  propagate f ctx (TypeSpec ts cant) = propagateTypeSpec f ctx ts cant
+  propagate f ctx (MemberSpec xs) = propagateMemberSpec f ctx xs
+
+-- ===========================
+-- Pred
+-- ===========================
+
+data Pred where
+  ElemPred :: forall a. HasSpec a => Bool -> Term a -> NonEmpty a -> Pred
+  And :: [Pred] -> Pred
+  Exists :: ((forall b. Term b -> b) -> GE a) -> Binder a -> Pred
+  ForAll :: (Container t a, HasSpec t, HasSpec a) => Term t -> Binder a -> Pred
+  DependsOn :: (HasSpec a, HasSpec b) => Term a -> Term b -> Pred
+  Assert :: Term Bool -> Pred
+  TruePred :: Pred
+  FalsePred :: NonEmpty String -> Pred
+  Case :: HasSpec (Either a b) => Term (Either a b) -> Binder a -> Binder b -> Pred
+  Let :: Term a -> Binder a -> Pred
+  Subst :: HasSpec a => Var a -> Term a -> Pred -> Pred
+
+data Binder a where
+  (:->) :: HasSpec a => Var a -> Pred -> Binder a
+
+class Container t e | t -> e where
+  fromForAllSpec :: (HasSpec t, HasSpec e) => Spec e -> Spec t
+  forAllToList :: t -> [e]
```

Comment:
In this code, do we use `forAll` on anything other than `Set`? In which case we could simplify this too so that `forAll` only supports `Set` and we could explain that you can easily generalize it if you want in text.

Reply:
We could do that. Or we could make it clear, that in the Minimal implementation it only is used on Set, but in the full implementation it is used in List and Map as well.